Add drmaa (for Python)

jakirkham opened the pull request:
Adds the Python DRMAA bindings package. Basically built using `conda skeleton pypi` with some tweaks to make it fit nicely in conda-forge.

* [x] Needs some notes explaining how to configure after installing.
    * [x] Also printed to the terminal when installed.
    * [x] Ideally, this will be brief, but clear.

jakirkham commented on recipes/drmaa/run_test.py:
```diff
@@ -0,0 +1,4 @@
+try:
+    import drmaa
```
Can't actually import as we need to set `DRMAA_LIBRARY_PATH` to an install of `libdrmaa.so`. Also would need an install of SGE and have `SGE_ROOT` and `SGE_CELL` set. So, instead we just verify that we can import, but will get the expected exception. This verifies it is doing what it should.

More sophisticated testing is unnecessary as `drmaa` is pure Python and goes through CI anyways.

dan-blanchard replied:
Yeah, that's the same problem I had when developing mine own recipe for it.

jjhelmus replied:
You could do the import test in a `run_test.sh` file where you set the DRMAA_LIBRARY_PATH and then run `python -c "import drmaa"`

jakirkham replied:
I think you will get a different error if it doesn't point to a real library. If it does point to a real library, but the wrong one, there will be errors about missing symbols and such.

We could check these I suppose. Just not sure if that is where we want to go here.

jakirkham replied:
If we really want to test it, we could install libdrmaa and sge with `yum` in the docker image. However, that takes awhile.

I don't know that we want to include that in the docker image by default. Though I do have a docker image for CentOS 6.6, which we could use here. It is setup a bit differently than the one used here and am not sure how much I want to change it for this purpose. However, we could construct a variant of that image for this purpose.

Thoughts?

jjhelmus replied:
Should a conda package for libdrmaa be created and this package depend on it?  If a user installed this package without libdrmaa installed on their system it wouldn't work would it.

dan-blanchard replied:
So the weird thing about libdrmaa is it's not actually a separate library, it's a library that each DRMAA-compatible batch system provides on their own.  So, installing SGE, Torque, PBS, etc. gets you their own versions of libdrmaa.

jakirkham replied:
Even if we include libdrmaa, it still won't quite work. This would require a fully configured and working SGE Cluster install.

We could do this for testing with the package manager. We could also use a docker container like this [one]( https://github.com/jakirkham/docker_centos_drmaa_conda ).

Though I agree with @dan-blanchard that taking SGE into `conda` feels like quite a bit of work. I looked at it once, but getting it built is intimidating to say the least and requires a host of dependencies that we don't have yet. Not to mention, it probably won't be configured correctly (or maybe work?) for most use cases.

A better solution might be to just add some notes here explaining clearly how this must be configured in order to get it to work. We could even have this printed during install using a pre or post link script.

Thoughts?

jjhelmus replied:
Makes sense.  I was not familiar with the details of libdrmaa, given the complexities I agree testing does not make sense.  Adding notes how to configure the package seems the best option.

jakirkham replied:
Sorry, didn't see that @dan-blanchard had already responded when I posted that.

jakirkham commented:
@dan-blanchard, do you have any thoughts on this recipe?

Also, maintenance and deployment of packages will primarily be occurring through conda-forge going forward as [noted]( https://github.com/conda-forge/conda-forge.github.io/issues/22#issuecomment-197117472 ). Some auditing will be performed before pulling packages from the `conda-forge` channel into the `defaults` channel.

Would you (and/or others) like to be added as a maintainer for `drmaa`? This will let you deploy releases to `conda-forge` automatically with a mere version bump in the recipe.

jakirkham commented:
[HTTP error]( https://travis-ci.org/conda-forge/staged-recipes/jobs/116950963#L481 ) during build. Will retrigger.

dan-blanchard commented:
@jakirkham, the recipe seems reasonable to me.  Like you mention, it's tricky to test because you need SGE installed, and making a conda recipe for that would be kind of crazy.

You can see the basic recipe I threw together a couple years ago [here](https://github.com/dan-blanchard/conda-recipes/blob/master/drmaa/meta.yaml).

I actually had recipes for quite a few packages that I never contributed to conda-recipes in my fork, but I'm sure many of them are out of date by now.

You can certainly put me down as maintainer for this, since it's not a package that gets updated very often.

jakirkham commented:
> ...the recipe seems reasonable to me.... You can see the basic recipe I threw together a couple years ago [here]( https://github.com/dan-blanchard/conda-recipes/blob/master/drmaa/meta.yaml ).

Good to hear. Looks like we are doing similar things.

> Like you mention, it's tricky to test because you need SGE installed, and making a conda recipe for that would be kind of crazy.

Yeah, definitely crazy. Can't say I didn't think about it once. :stuck_out_tongue_winking_eye:

I did add SGE to a docker container though, but that doesn't feel like what we should be doing here.

> I actually had recipes for quite a few packages that I never contributed to conda-recipes in my fork, but I'm sure many of them are out of date by now.

If you are willing to contribute them, that would be awesome. Probably could help with the effort if you could provide a short list of what didn't get contributed and we could go through adding them here. Would help you get more mileage out of them at least.

> You can certainly put me down as maintainer for this, since it's not a package that gets updated very often.

Excellent. Thanks for stepping up. I added you. Once this gets merged (and a few other automated things happen), you will be invited to join conda-forge and, in particular, to a team for managing `drmaa`.

pelson commented:
Worrying that the test passes, but it doesn't look like anything is installed on Windows (I didn't see the bld.bat).

This has gone green, are you getting to the point of wanting a feedstock to iterate in?

jakirkham commented:
> Worrying that the test passes, but it doesn't look like anything is installed on Windows (I didn't see the bld.bat).

For all practical purposes this could be restricted to Linux.

Normally this is only used on cluster where someone has access to some sort of job scheduler like SGE, PBS, etc. Almost always that is Linux, but I did see someone try to build SGE on OS X. It might be possible to do something similar on Windows. Though honestly getting any of these set up and working without a package manager feels like a full time job. :wink:

If you think it should be restricted to Linux, that would be ok by me. I could also add a `bld.bat` for the brave soul running SGE or other on it.

> This has gone green, are you getting to the point of wanting a feedstock to iterate in?

I probably should add a prelink script that echos a brief warning/setup instructions before we merge though.

pelson commented:
Why not go ahead and skip anything that isn't linux. We can always change that in the future.

jakirkham commented:
I will update it.

pelson commented:
Do we also need to change the name of the package?

pelson commented:
I'm inclined to suggest drmaa-python...

jakirkham commented:
Hmm...so it is called `drmaa` on [PyPI]( https://pypi.python.org/pypi/drmaa ) and in the [`anaconda` channel]( https://anaconda.org/anaconda/drmaa ). This would be a break from both of those.

jakirkham commented:
Not to mention, we basically will never package libdrmaa as it would require building a job scheduler like SGE (too much work for no real benefit). Even if we did, I would be inclined to call it `libdrmaa` and add some features to specify the variants.

pelson commented:
> Hmm...so it is called drmaa on PyPI and in the anaconda channel. This would be a break from both of those.

I accept the anaconda argument (though not the PyPI one :wink:). I'm happy to be consistent with anaconda where the choice isn't clearly damaging.

Thanks for putting together a solid case for keeping it as ```drmaa```. I agree :+1:

jakirkham commented:
I definitely understand your point about PyPI packages. The `pylibtiff` package renamed itself to [`libtiff`]( https://pypi.python.org/pypi/libtiff/ ) at some point, which always drove me nuts. Especially when placing it in `conda`.

pelson commented:
What do you want to do with this one @jakirkham - there are still a number of boxes which are left unchecked and it is still called WIP...

conda-forge-admin commented:
Hi! This is the friendly conda-forge-admin automated user.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipes/python-drmaa```) and found it was in an excellent condition.

pelson commented on recipes/python-drmaa/meta.yaml:
```diff
@@ -0,0 +1,38 @@
+
+test:
+  files:
+    - run_test.py
```
This should no longer be needed (the linter knows about run_test.py now)

jakirkham replied:
Ah, ok.

jakirkham replied:
Is it ok if I leave it though? I kind of like the visibility we are giving it in the `meta.yaml`. If you don't like it still, I can of course remove it.

pelson replied:
It is fine 👍

jakirkham replied:
Awesome. Thanks.

If you change your mind, I will be happy to address it. Just feeling guilty for letting this sit for like a month. :disappointed:

jakirkham commented:
Added a post link script message. Renamed to `python-drmaa` to avoid potential confusion. Fixed the license version as it was unclear. This should be ready as long as CI passes.

jakirkham commented:
This seems to be passing. Any final thoughts? Kind of eager to get this out of the backlog. :grinning:

jakirkham commented:
Alright, I'm getting this one out. If there are any concerns, please raise them at the feedstock.

jakirkham commented:
I should add that while Continuum does call this `drmaa`. I have opted to break that convention here and call it `python-drmaa`. My reasoning is simple.

DRMAA is a specification and a C library provided with cluster queue installs. There are many language bindings to that specification. As we may be including these in the future, I don't want this one to be any different in this regard. This is merely the Python implementation of bindings to DRMAA. Further, calling it `drmaa` may give the user a false idea as to what this is used for. It does not install the DRMAA library for a cluster. This remains up to the user to do (or more like a system administrator).

I know breaks like these can be controversial. However, I hope this rationale sounds reasonable. I am happy to discuss this with others. Also, am happy to work with others to address any concerns this may raise.

pelson commented:
> I know breaks like these can be controversial. However, I hope this rationale sounds reasonable

I certainly support it.

Here is a really controversial idea for the linter:

```
if 'python' in meta.ms_depends('run') and 'py' not in meta.name():
    lint.append('A python dependent package should have a "py" somewhere in the name')
```

I say this only in semi-jest 😉

jakirkham commented:
> I certainly support it.

Well that certainly brings down by stress level. :relieved:

> Here is a really controversial idea for the linter

Yeah, it is really hard to say in some cases. For instance, I don't know of any non-Python packages called `webcolors`, but that could always change, right?

Still I saw someone adding a Python package called `hdfs` the other day and immediately had to suggest a `python-` prefix. It would almost be nice for the linter to say something here. So, if we forget, it comes up at least once (even if it might be wrong).

Another example is the `protobuf` package here was Python only previously (just as the `defaults` one is), but `protobuf` is originally C++ with other language support added later. So, I have shoved C++ support into it so it would behave as people would expect it to (like having `protoc`). This made the versioning a little odd too as the Python package had post releases that were not included in the repo. Though if we do that with other languages it supports (Java, Ruby, etc.) that package is going to be huge. We almost would be better off to split it into smaller packages. However, I think we should probably wait in that case and see how it goes.

dan-blanchard commented:
I've actually complained for a long time that conda packages should all have a language prefix.  If it's a going to be a language-agnostic package manager, the python packages should start with `py-` or `python-`.  I actually added the support for creating perl packages with `conda skeleton` a couple years ago, and I immediately realized that this was going to be a big problem as support was added for additional languages.
